README.md
Dart wrapper via `dart:js` for https://webbluetoothcg.github.io/web-bluetooth/

## Features

- canUseBlue
- getAvailability
- subscribeAvailabilitychanged/unsubscribeAvailabilitychanged
- getDevices/requestDevice
- connect/disconnect
- getPrimaryService/getCharacteristic
- startNotifications/stopNotifications
- subscribeValueChanged/unsubscribeValueChanged
- readValue/writeValueWithResponse

## Usage

### canUseBlue

```dart
bool canUse = canUseBlue();
print('canUse $canUse');
```

### getAvailability

https://developer.mozilla.org/en-US/docs/Web/API/Bluetooth/getAvailability

```dart
bool availability = await blue.getAvailability();
print('availability $availability');
```

### subscribeAvailabilitychanged/unsubscribeAvailabilitychanged

https://developer.mozilla.org/en-US/docs/Web/API/Bluetooth/onavailabilitychanged

```dart
final EventListener _handleAvailabilitychanged = allowInterop((Event event) {}
...
blue.subscribeAvailabilitychanged(_handleAvailabilitychanged);
...
blue.unsubscribeAvailabilitychanged(_handleAvailabilitychanged);
```

### getDevices/requestDevice

- https://developer.mozilla.org/en-US/docs/Web/API/Bluetooth/getDevices

```dart
List<BlueDevice> getDevices = await blue.getDevices();
_device = getDevices[0];
```

- https://developer.mozilla.org/en-US/docs/Web/API/Bluetooth/requestDevice

```dart
BlueDevice requestDevice = await blue.requestDevice(RequestOptions(
  optionalServices: [BlueUUID.getService(WOODEMI_SERV__COMMAND)],
  acceptAllDevices: true,
));
_device = requestDevice;
```

### connect/disconnect

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTServer/connect

```dart
_device?.gatt.connect().then((value) {
  print('device.gatt.connect success');
}).catchError((error) {
  print('device.gatt.connect $error');
});
```

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTServer/disconnect

```dart
_device?.gatt.disconnect();
```

### getPrimaryService/getCharacteristic

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTServer/getPrimaryService

```dart
_service = await _device!.gatt.getPrimaryService(BlueUUID.getService(WOODEMI_SERV__COMMAND));
```

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTService/getCharacteristic

```dart
_characteristic = await _service!.getCharacteristic(BlueUUID.getCharacteristic(WOODEMI_CHAR__COMMAND_REQUEST));
```

### startNotifications/stopNotifications

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTCharacteristic/startNotifications

```dart
await _characteristic!.startNotifications();
```

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTCharacteristic/stopNotifications

```dart
await _characteristic!.stopNotifications();
```

### subscribeValueChanged/unsubscribeValueChanged

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTCharacteristic/characteristicvaluechanged

```dart
final EventListener _handleValueChanged = allowInterop((event) {}
...
_characteristic!.subscribeValueChanged(_handleValueChanged);
...
_characteristic!.unsubscribeValueChanged(_handleValueChanged);
```

### readValue/writeValueWithResponse

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTCharacteristic/readValue

```dart
var byteData = await _characteristic!.readValue();
```

- https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTCharacteristic/writeValueWithResponse

```dart
_characteristic!.writeValueWithResponse(Uint8List.fromList([0x01, 0x0A, 0x00, 0x00, 0x00, 0x01]));
```

## Additional information

Status in Chromium: https://chromestatus.com/feature/5264933985976320


CHANGELOG.md
# 0.2.0

- subscribeAvailabilitychanged/unsubscribeAvailabilitychanged
- getDevices/requestDevice
- connect/disconnect
- getPrimaryService/getCharacteristic
- startNotifications/stopNotifications
- subscribeValueChanged/unsubscribeValueChanged
- readValue/writeValueWithResponse

## 0.1.0

- canUseBlue
- getAvailability


test/web_blue_test.dart
import 'package:web_blue/web_blue.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    final awesome = Awesome();

    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {
      expect(awesome.isAwesome, isTrue);
    });
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:web_blue_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


example/lib/main.dart
// ignore_for_file: avoid_print

import 'package:flutter/material.dart';
import 'package:web_blue/web_blue.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('web_blue example'),
      ),
      body: Center(
        child: Column(
          children: [
            ElevatedButton(
              child: const Text('canUse'),
              onPressed: () {
                bool canUse = canUseBlue();
                print('canUse $canUse');
              },
            ),
            ElevatedButton(
              child: const Text('getAvailability'),
              onPressed: () async {
                bool availability = await blue.getAvailability();
                print('availability $availability');
              },
            ),
            ElevatedButton(
              child: const Text('BlueUUID.getService'),
              onPressed: () async {
                BlueUUID.getService('device_information');
              },
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/web_blue_base.dart
part of '../web_blue.dart';

class Blue extends InteropWrapper<_Blue> {
  final _Blue _interop;

  Blue._(this._interop);

  Future<bool> getAvailability() {
    var promise = _interop.getAvailability();
    return promiseToFuture(promise);
  }

  // FIXME allowInterop
  void subscribeAvailabilitychanged(EventListener listener) {
    _interop.addEventListener('availabilitychanged', listener);
  }

  // FIXME allowInterop
  void unsubscribeAvailabilitychanged(EventListener listener) {
    _interop.removeEventListener('availabilitychanged', listener);
  }

  Future<BlueDevice> requestDevice([RequestOptions? options]) {
    var promise = _interop.requestDevice(options);
    return promiseToFuture(promise).then((value) => BlueDevice._(value));
  }

  Future<List<BlueDevice>> getDevices() {
    var promise = _interop.getDevices();
    return promiseToFuture(promise).then((value) {
      return (value as List).map((e) => BlueDevice._(e)).toList();
    });
  }
}

@JS('Bluetooth')
class _Blue implements Interop {
  /// https://developer.mozilla.org/en-US/docs/Web/API/Bluetooth/getAvailability
  external Object getAvailability();

  /// https://developer.mozilla.org/en-US/docs/Web/API/Bluetooth/requestDevice
  external Object requestDevice(Object? options);

  /// https://developer.mozilla.org/en-US/docs/Web/API/Bluetooth/getDevices
  external Object getDevices();

  /// https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener
  external void addEventListener(String type, EventListener? listener, [bool? useCapture]);

  /// https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener
  external void removeEventListener(String type, EventListener? listener, [bool? useCapture]);
}

@JS()
@anonymous
class RequestOptions {
  external factory RequestOptions({
    List<RequestOptionsFilter>? filters,
    List<Object>? optionalServices,
    bool? acceptAllDevices,
  });
}

@JS()
@anonymous
class RequestOptionsFilter {
  external factory RequestOptionsFilter({
    List<Object>? services,
  });
}

@JS('BluetoothUUID.getService')
external Object _getService(String name);

@JS('BluetoothUUID.getCharacteristic')
external Object _getCharacteristic(String name);

class BlueUUID {
  BlueUUID._();

  static Object getService(String name) => _getService(name);

  static Object getCharacteristic(String name) => _getCharacteristic(name);
}

class BlueDevice extends Delegate<Object> {
  final BlueRemoteGATTServer gatt;

  BlueDevice._(Object delegate)
      : gatt = BlueRemoteGATTServer._(getProperty(delegate, 'gatt')),
        super(delegate);
}

class BlueRemoteGATTServer extends Delegate<Object> {
  BlueRemoteGATTServer._(Object delegate) : super(delegate);

  Future<BlueRemoteGATTServer> connect() {
    var promise = callMethod('connect');
    return promiseToFuture(promise).then((_) => this);
  }

  void disconnect() => callMethod('disconnect');

  Future<BlueRemoteGATTService> getPrimaryService(Object bluetoothUuid) {
    var promise = callMethod('getPrimaryService', [bluetoothUuid]);
    return promiseToFuture(promise).then((result) => BlueRemoteGATTService._(result));
  }
}

class BlueRemoteGATTService extends Delegate<Object> {
  BlueRemoteGATTService._(Object delegate) : super(delegate);

  Future<BlueRemoteGATTCharacteristic> getCharacteristic(Object bluetoothUuid) {
    var promise = callMethod('getCharacteristic', [bluetoothUuid]);
    return promiseToFuture(promise).then((result) => BlueRemoteGATTCharacteristic(result));
  }
}

class BlueRemoteGATTCharacteristic extends Delegate<EventTarget> {
  BlueRemoteGATTCharacteristic(EventTarget delegate) : super(delegate);

  String get uuid => getPropertyT('uuid');

  ByteData get value => getPropertyT('value');

  Future<BlueRemoteGATTCharacteristic> startNotifications() {
    var promise = callMethod('startNotifications');
    return promiseToFuture(promise).then((_) => this);
  }

  Future<BlueRemoteGATTCharacteristic> stopNotifications() {
    var promise = callMethod('stopNotifications');
    return promiseToFuture(promise).then((_) => this);
  }

  // FIXME allowInterop
  void subscribeValueChanged(EventListener listener) {
    delegate.addEventListener('characteristicvaluechanged', listener);
  }

  // FIXME allowInterop
  void unsubscribeValueChanged(EventListener listener) {
    delegate.removeEventListener('characteristicvaluechanged', listener);
  }

  Future<ByteData> readValue() {
    var promise = callMethod('readValue');
    return promiseToFuture(promise);
  }

  Future<void> writeValueWithResponse(TypedData bytes) {
    var promise = callMethod('writeValueWithResponse', [bytes]);
    return promiseToFuture(promise);
  }

  Future<void> writeValueWithoutResponse(TypedData bytes) {
    var promise = callMethod('writeValueWithoutResponse', [bytes]);
    return promiseToFuture(promise);
  }
}

example/lib/woodemi_notepad_page.dart
// ignore_for_file: avoid_print, avoid_web_libraries_in_flutter

import 'dart:html' show EventListener;
import 'dart:js' show allowInterop;
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:web_blue/web_blue.dart';

const WOODEMI_SUFFIX = 'ba5e-f4ee-5ca1-eb1e5e4b1ce0';

const WOODEMI_SERV__COMMAND = '57444d01-$WOODEMI_SUFFIX';
const WOODEMI_CHAR__COMMAND_REQUEST = '57444e02-$WOODEMI_SUFFIX';
const WOODEMI_CHAR__COMMAND_RESPONSE = WOODEMI_CHAR__COMMAND_REQUEST;

class WoodemiNotepadPage extends StatefulWidget {
  const WoodemiNotepadPage({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() => _WoodemiNotepadPageState();
}

class _WoodemiNotepadPageState extends State<WoodemiNotepadPage> {
  BlueDevice? _device;
  BlueRemoteGATTService? _service;
  BlueRemoteGATTCharacteristic? _characteristic;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Woodemi Notepad'),
      ),
      body: Center(
        child: Column(
          children: [
            _buildRequestGet(),
            _buildConnectDisconnect(),
            _buildServiceCharacteristic(),
            _buildStartStopNotifications(),
            _buildValueChangedSubscription(),
            _buildReadWriteValue(),
          ],
        ),
      ),
    );
  }

  Widget _buildRequestGet() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        ElevatedButton(
          child: const Text('requestDevice'),
          onPressed: () async {
            BlueDevice requestDevice = await blue.requestDevice(RequestOptions(
              optionalServices: [BlueUUID.getService(WOODEMI_SERV__COMMAND)],
              acceptAllDevices: true,
            ));
            print('requestDevice $requestDevice');
            _device = requestDevice;
          },
        ),
        ElevatedButton(
          child: const Text('getDevices'),
          onPressed: () async {
            List<BlueDevice> getDevices = await blue.getDevices();
            print('getDevices $getDevices');
            _device = getDevices[0];
          },
        ),
      ],
    );
  }

  Widget _buildConnectDisconnect() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        ElevatedButton(
          child: const Text('device.gatt.connect'),
          onPressed: () {
            _device?.gatt.connect().then((value) {
              print('device.gatt.connect success');
            }).catchError((error) {
              print('device.gatt.connect $error');
            });
          },
        ),
        ElevatedButton(
          child: const Text('device.gatt.disconnect'),
          onPressed: () async {
            _device?.gatt.disconnect();
            print('device.gatt.disconnect success');
          },
        ),
      ],
    );
  }

  Widget _buildServiceCharacteristic() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        ElevatedButton(
          child: const Text('device.getPrimaryService'),
          onPressed: () async {
            _service = await _device!.gatt.getPrimaryService(BlueUUID.getService(WOODEMI_SERV__COMMAND));
            print('service $_service');
          },
        ),
        ElevatedButton(
          child: const Text('service.getCharacteristic'),
          onPressed: () async {
            _characteristic = await _service!.getCharacteristic(BlueUUID.getCharacteristic(WOODEMI_CHAR__COMMAND_REQUEST));
            print('characteristic $_characteristic');
          },
        ),
      ],
    );
  }


  Widget _buildStartStopNotifications() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        ElevatedButton(
          child: const Text('characteristic.startNotifications'),
          onPressed: () async {
            await _characteristic!.startNotifications();
            print('characteristic.startNotifications success');
          },
        ),
        ElevatedButton(
          child: const Text('characteristic.stopNotifications'),
          onPressed: () async {
            await _characteristic!.stopNotifications();
            print('characteristic.stopNotifications success');
          },
        ),
      ],
    );
  }

  Widget _buildValueChangedSubscription() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        ElevatedButton(
          child: const Text('characteristic.subscribeValueChanged'),
          onPressed: () {
            _characteristic!.subscribeValueChanged(_handleValueChanged);
            print('characteristic.subscribeValueChanged success');
          },
        ),
        ElevatedButton(
          child: const Text('characteristic.unsubscribeValueChanged'),
          onPressed: () {
            _characteristic!.unsubscribeValueChanged(_handleValueChanged);
            print('characteristic.unsubscribeValueChanged success');
          },
        ),
      ],
    );
  }

  Widget _buildReadWriteValue() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        ElevatedButton(
          child: const Text('characteristic.readValue'),
          onPressed: () async {
            var byteData = await _characteristic!.readValue();
            print('characteristic.readValue $byteData');
          },
        ),
        ElevatedButton(
          child: const Text('characteristic.writeValueWithResponse'),
          onPressed: () async {
            _characteristic!.writeValueWithResponse(Uint8List.fromList([0x01, 0x0A, 0x00, 0x00, 0x00, 0x01]));
            print('characteristic.writeValueWithResponse success');
          },
        ),
      ],
    );
  }
}

final EventListener _handleValueChanged = allowInterop((event) {
  var characteristic = BlueRemoteGATTCharacteristic(event.target!);
  print('_handleValueChanged ${characteristic.uuid}, ${characteristic.value}');
});
